### PR TITLE
make trail history persistent (fix #2958)

### LINKED ISSUE #2958: Have trail on live map more permanent
From FB:
When one exits the map and enter the cachedescription for logging and then enters the map afterwards the trace is gone.

Request: Track on map log. Would like a track log that shows the track even if one have closed the map and opened it once again. Just like a track log on a real GPS.

Reasson: Have been some places where mapcoverage isn't good and would like the possibilities to "backtrace" ones footsteps.
## 

Also when leaving the live map by using the back key the trail is lost.
It would be could to keep it running/saved as long as c:geo is active with a menu option on the map menu to clear it.

### PR DESCRIPTION
**Pull request by moving-bits**
- log trail history to database
- restore when recreating map (implemented for CGeoMap and NewMap)
- menu option to clear history (implemented for CGeoMap and NewMap)
- limit number of trail points in database automatically (cleanup on open db)

### CODE REVIEW

**Comment by moving-bits** [comment]
Could someone have a look at my function that writes the data to the db in the background, please? `saveToStorage()` in PositionHistory.java#34 - It seems to work, but I'm not sure if it's the right way to do. Thanks.

**Comment by rsudev** [comment]
As I just came across the bug that triggered this implementation, I took a short first look.
I would think it should be more efficient to save all track points to the database when we leave the map, not on every move, shouldn't it?
As the PositionHistory is self-initializing now, the Activity-state persisting should then no longer take it into account (but could trigger the save to db).

**Comment by moving-bits** [comment]
ok, will try to adapt my PR after you've checked in the bug fix for #7235

If I understood it correctly, I would:
- change the place where saving gets triggered + save the whole trail (instead of just a single point) up to a certain maximum length
- therefore I do not need to strip the saved trail on DB startup anymore, since max length is limited on save
- keep the DB structure change (new table) as is in the PR
- probably have to change the place where the trail gets loaded (initializing of PositionHistory instead of Activity init)

**Comment by kumy** [comment]
Testing jenkins - related to #7295 

Jenkins: retest this please

**Comment by kumy** [comment]
Jenkins: retest this please

**Comment by Lineflyer** [comment]
retest this please

**Comment by Lineflyer** [comment]
I suggest to postpone merging this new feature until we have "stable" base for the next release and have branched to `release` branch.

**Comment by moving-bits** [comment]
Yes - this PR was not supposed to be merged in a situation like today. I think kumy used it just for testing purposes today.

**Comment by Lineflyer** [comment]
Setting "Do not merge" for the time being to have better overview.

**Comment by moving-bits** [comment]
Hm, this wasn't supposed to be merged right now. Need to talk to @rsudev first to get a grip on how to implement the changes he suggested. Readiness to merge won't be until next week, probably.

**Comment by Lineflyer** [comment]
I did not merge it, just removed the label. But I added the label again and changed the title to make sure no one merges yet :)

**Comment by moving-bits** [comment]
Maybe it was the CI "waiting for input" and taking the removal of the "Do not merge" label as call to action :)
Thanks for re-placing it.

**Comment by Lineflyer** [comment]
> Maybe it was the CI "waiting for input" and taking the removal of the "Do not merge" label as call to action :)

No, it was as I pushed a commit to your branch for this PR (Typo fix).

**Comment by moving-bits** [comment]
> No, it was as I pushed a commit to your branch for this PR (Typo fix).

ah, I see

**Comment by moving-bits** [comment]
> As I just came across the bug that triggered this implementation, I took a short first look.
> I would think it should be more efficient to save all track points to the database when we leave the map, not on every move, shouldn't it?
> As the PositionHistory is self-initializing now, the Activity-state persisting should then no longer take it into account (but could trigger the save to db).

@rsudev: Could you give me a hint or two on:
- where to place the saving of the trailhistory to db after your changes?
- whether loading of the trailhistory from db can stay where it is now?
- whether the way I initiate the db write in background is technically ok?

Thanks!

File: main/src/cgeo/geocaching/maps/PositionHistory.java [hunk 1]
```diff
@@ -21,6 +25,26 @@
 
     private ArrayList<Location> history = new ArrayList<>();
 
+    // load data from permanent storage
+    public PositionHistory() {
```
Comment (rsudev) [inline main/src/cgeo/geocaching/maps/PositionHistory.java]: PositionHistory is only used on CGeoMap if I am not mistaken, so the persistence will only work there, or did I miss something?

File: main/src/cgeo/geocaching/maps/PositionHistory.java [hunk 2]
```diff
@@ -41,6 +66,7 @@ public void rememberTrailPosition(final Location coordinates) {
             return;
         }
 
+        saveToStorage(coordinates);
```
Comment (rsudev) [inline main/src/cgeo/geocaching/maps/PositionHistory.java]: Essentially I would prefer to save the history only when the map activity is stopped, but when I looked through the implementation, it seemed not too easy to get that triggered in a safe fashion, so we might leave it like it is implemented here for the moment.

**Comment by Lineflyer** [comment]
@moving-bits Can you double confirm you consider this change ready to merge?`We just copied `master` to `release`so its the best time for new feature merging onto `master`.

**Comment by moving-bits** [comment]
rebased, retested, squashed and pushed => feel free to merge this one

**Pull request merged**